vslib/src/SwitchStateBase.cpp

```diff
@@ -2424,3 +2438,183 @@ void SwitchStateBase::debugSetStats(
+                {
+                    SWSS_LOG_ERROR("Invalid voq system port config info! sys port count %d, sys port list %p", sys_port_count, sys_port_cfg_list);
+                    return (int32_t)SAI_STATUS_INVALID_ATTR_VALUE_0 + (int32_t)i;
+                }
+                break;
+
+            default:
+                // Ignore other attributes. They are processed elsewhere.
+                break;
+        }
+    }
+
+    if (!voq_switch)
+    {
+        // No switch type attribute in the attribute list.
+        return SAI_STATUS_SUCCESS;
+    }
+
+    CHECK_STATUS(create_system_ports(voq_switch_id, sys_port_count, sys_port_cfg_list));
+
+    CHECK_STATUS(set_system_port_list());
+
+    CHECK_STATUS(set_voq_switch_attributes(voq_switch_id, voq_max_cores));
+
+    return SAI_STATUS_SUCCESS;
+}
+
+sai_status_t SwitchStateBase::create_system_ports(
+       _In_ int32_t voq_switch_id,
+       _In_ uint32_t sys_port_count,
+       _In_ const sai_system_port_config_t *sys_port_cfg_list)
+{
+    SWSS_LOG_ENTER();
+
+    SWSS_LOG_INFO("create system ports");
+
+    if(!sys_port_cfg_list)
+    {
+        return SAI_STATUS_FAILURE;
+    }
+
+    m_system_port_list.clear();
+
+    for (uint32_t i = 0; i < sys_port_count; i++)
+    {
+        SWSS_LOG_DEBUG("create system port index %u", i);
+
+        sai_object_id_t system_port_id;
+
+        CHECK_STATUS(create(SAI_OBJECT_TYPE_SYSTEM_PORT, &system_port_id, m_switch_id, 0, NULL));
+
+        m_system_port_list.push_back(system_port_id);
+
+        sai_attribute_t attr;
+
+        attr.id = SAI_SYSTEM_PORT_ATTR_CONFIG_INFO;
+        attr.value.sysportconfig = sys_port_cfg_list[i];
+
+        CHECK_STATUS(set(SAI_OBJECT_TYPE_SYSTEM_PORT, system_port_id, &attr));
+
+        attr.id = SAI_SYSTEM_PORT_ATTR_TYPE;
+        attr.value.s32 = SAI_SYSTEM_PORT_TYPE_REMOTE;
+
+        if ((int32_t)sys_port_cfg_list[i].attached_switch_id == voq_switch_id)
+        {
+            attr.value.s32 = SAI_SYSTEM_PORT_TYPE_LOCAL;
+
+            // Need to set local port oid attribute TODO
+        }
+
+        CHECK_STATUS(set(SAI_OBJECT_TYPE_SYSTEM_PORT, system_port_id, &attr));
+    }
+
+    return SAI_STATUS_SUCCESS;
+}
+
+sai_status_t SwitchStateBase::set_system_port_list()
+{
+    SWSS_LOG_ENTER();
+
+    SWSS_LOG_INFO("set system port list");
+
+    // NOTE: this is static, but will be refreshed on read only get
+
+    sai_attribute_t attr;
+
+    uint32_t sys_port_count = (uint32_t)m_system_port_list.size();
+
+    attr.id = SAI_SWITCH_ATTR_SYSTEM_PORT_LIST;
+    attr.value.objlist.count = sys_port_count;
+    attr.value.objlist.list = m_system_port_list.data();
+
+    CHECK_STATUS(set(SAI_OBJECT_TYPE_SWITCH, m_switch_id, &attr));
+
+    attr.id = SAI_SWITCH_ATTR_NUMBER_OF_SYSTEM_PORTS;
+    attr.value.u32 = sys_port_count;
+
+    return set(SAI_OBJECT_TYPE_SWITCH, m_switch_id, &attr);
+}
+
+sai_status_t SwitchStateBase::set_voq_switch_attributes(
```

Comment:
as for this function, set of those 2 attributes is not required
https://github.com/opencomputeproject/SAI/blob/master/inc/saiswitch.h#L2175 - you already check that in initialize_voq_objects
also SAI_SWITCH_ATTR_TYPE is checked to be equal to VOQ in initialize objects
https://github.com/opencomputeproject/SAI/blob/master/inc/saiswitch.h#L2184 also checked in initialize objects

those 3 attributes you SET here, they are already passed on attr_list when calling switch_create, so there is not need for this entire function at all

Reply:
I want values of these three attributes or at least swith_id and max_cores in the objectHash for retrieval later (by GET from orchagent, even though not doing now, we'll need it in the future) and to emulate the SAI as close as possible to real SAI. If I do not "set" in this function, will they be in objectHash? Would you please let me know where they are set?

Reply:
they will be posted into objectHash after init_switch, and orchagent will be able to get them each api call is under mutex so no issue here,
objectHash is set in VirtualSwitchSaiInterface::create(
675     return ss->create(object_type, serializedObjectId, switchId, attr_count, attr_list);

Reply:
o.k. Thanks. I'll fix as proposed

Reply:
Fixed